### ISSUE TITLE
Hessian should be zero/Jacobian coord?

### ISSUE THREAD

**Issue by abelsiqueira** [JuliaSmoothOptimizers/NLPModelsJuMP.jl#13]
Long post, sorry. I'm having a very specific issue related to JuliaSmoothOptimizers/NLPModels.jl#146 these lines

https://github.com/JuliaSmoothOptimizers/NLPModelsJuMP.jl/blob/2613241e031e79161e054e1f76d60100af1c57e7/src/nlp_to_mpb.jl#L47-L56

Essentially, my model computes `hess_coord` with an error that sometimes vanishes. When creating the model, the structure is stored with 4 nonzeros, but during execution `hess_coord` returns 5 nonzeros. The problem is that the scruture has 4 elements and `hess_coord` now wants to store 5.

Now with more details. I have the NLSModel [mgh04](https://github.com/JuliaSmoothOptimizers/NLSProblems.jl/blob/master/src/mgh04.jl) and the [`NLS to constrained LS`](https://github.com/JuliaSmoothOptimizers/NLPModels.jl/pull/146/files#diff-bd75dc0ee43b933a25361b0aba1e563eR170) model. The Hessian of `F3(x) = x[1] * x[2] - 2` should be `[0 1; 1 0]`, but any of these zeros can turn into an error of order `eps()`.
This can be recreated without NLS:
```
using JuMP, NLPModels, NLPModelsJuMP
model = Model()
@variable(model, x[1:2])
@NLobjective(model, Min, 0.0)
@NLconstraint(model, x[1] * x[2] == 2)
nlp = MathProgNLPModel(model)
hess_coord(nlp, [0.659; 0.702], y=[1.0])
```
the output is `([1, 2, 2], [1, 2, 1], [0.0, 1.11022e-16, 1.0])`, which means some of the elements that should be 0 is actually close to `1e-16`. Now, this doesn't produce an error, because the structure is such that these "null" elements are stored.

However, when I convert a `MathProgNLSModel` using `NLStoCons`, it loses that information and returns only the nonzero elements. This becomes a problem when I pass it to `NLPtoMPB` as explained above.

Now, a possibility of workaround is to change the code for `eval_hesslag` in `nlp_to_mpb.jl` and return only the `vals` for which `rows` and `cols` are defined in the structure.

**Comment by dpo** [comment]
Is the `eps()` generated by JuMP/MPB or by `MathProgNLPModel`? It's odd that automatic differentiation would not produce exactly zero. Are you able to define a pure JuMP model and obtain the nonzero structure of the Hessian with just MPB (no `MathProgNLPModel` involved)?

**Comment by abelsiqueira** [comment]
Here:
```
using JuMP, MathProgBase
model = Model()
@variable(model, x[1:2])
@NLobjective(model, Min, 0.0)
@NLconstraint(model, x[1] * x[2] == 2)
d = JuMP.NLPEvaluator(model)
MathProgBase.initialize(d, [:Hess])
hx = zeros(3)
MathProgBase.eval_hesslag(d, hx, [0.659; 0.702], 1.0, [1.0])
```

**Comment by abelsiqueira** [comment]
Yes. The epsilon continues to appear on the MPB level so that doesn't get fixed, but mantaining `nnzh` fixed solves the initial problem.
